### PR TITLE
no longer create a default chain before execute every command

### PR DESCRIPTION
**Pull request by polunzh**

### CODE REVIEW

File: src/cli/chain/list/list.js [hunk 1]
```diff
@@ -36,6 +36,7 @@ const getChainStatusColor = status => {
 async function printChains() {
   const chains = getAllChainNames();
   if (chains.length === 0) {
+    printInfo(`There is no chains, you can create by run ${chalk.cyan('forge chain:create')}`);
```
Comment (wangshijun) [inline src/cli/chain/list/list.js]: there are no chains, you can create one by ...

File: src/cli/blocklet/use/index.js [hunk 2]
```diff
@@ -12,6 +12,8 @@ cli('blocklet:use', 'Download and install a blocklet', input => action(execute,
     runningNode: false,
     rpcClient: false,
     wallet: false,
+    chainName: false,
+    chainExists: false,
```
Comment (polunzh) [inline src/cli/blocklet/use/index.js:16]: 添加注释

File: src/cli/chain/create/index.js [hunk 3]
```diff
@@ -9,11 +9,16 @@ cli(
   input => action(execute, run, input),
   {
     alias: 'create-chain',
+    parseArgs: chainName => ({
```
Comment (polunzh) [inline src/cli/chain/create/index.js:12]: 注释

File: src/cli/node/install/install.js [hunk 4]
```diff
@@ -91,7 +95,7 @@ async function main({
         return;
       }
     }
-    print(`Now you can start a forge node with ${chalk.cyan('forge start')}`);
+    print(`Now you can create a forge node with ${chalk.cyan('forge chain:create')}`);
```
Comment (polunzh) [inline src/cli/node/install/install.js]: `Now you can create a chain...`

File: src/core/libs/common.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Common functions, different from `core/util`, this module is at higher level than `core/util`
+ */
+const { getAllChainNames, listReleases, updateReleaseYaml } = require('../forge-fs');
+
+const DEFAULT_CHAIN_NAME_RETURN = { no_chains: 1 };
```
Comment (polunzh) [inline src/core/libs/common.js]: 大写

File: src/cli/node/stop/index.js [hunk 6]
```diff
@@ -12,7 +15,22 @@ cli(
       forgeRelease: true,
       runningNode: false,
       rpcClient: false,
+      chainName: async ({ chainName }) => {
```
Comment (polunzh) [inline src/cli/node/stop/index.js]: 可以用`requirement.currentChainRunning`

File: src/cli/node/stop/index.js [hunk 7]
```diff
@@ -12,7 +15,22 @@ cli(
...
+
+        if (allProcesses.length === 0) {
+          printWarning('No running processes');
+          process.exit(0);
+        }
+
+        return allProcesses[0].name;
+      },
     },
+    parseArgs: chainName => chainName,
```
Comment (polunzh) [inline src/cli/node/stop/index.js]: bug: chainName => ({chainName})

File: src/cli/node/stop/stop.js [hunk 8]
```diff
@@ -77,6 +69,14 @@ async function main({ opts: { force, all }, args: [chainName = process.env.FORGE
     deprecated('forge stop --force: Use forge stop --all instead');
   }
 
+  const allProcesses = await getAllRunningProcesses();
```
Comment (polunzh) [inline src/cli/node/stop/stop.js]: 可以删除

File: src/cli/release/use/index.js [hunk 9]
```diff
@@ -13,14 +13,16 @@ cli(
...
+      chainExists: false,
     },
     options: [],
     handlers: {
       '--help': () => {
         shell.echo(`
 Examples:
-  - ${chalk.cyan('forge use 0.22.0')}      activate forge v0.22.0
-  - ${chalk.cyan('forge use v0.22.0')}     activate forge v0.22.0
+  - ${chalk.cyan('forge use 0.38.0')}      activate forge v0.38.0
```
Comment (polunzh) [inline src/cli/release/use/index.js]: 可以读 engines.forge
还有 install/index.js

File: src/core/env.js [hunk 10]
```diff
@@ -36,23 +43,32 @@ const config = { cli: {} }; // global shared forge-cli run time config
  * Setup running env for various commands, the check order for each requirement is important
  * Since a running node requires a release directory
  * An unlocked wallet requires a valid config
- *
+ * TODO: args 参数应该可以去掉
```
Comment (polunzh) [inline src/core/env.js]: args 参数应该可以去掉

File: src/core/env.js [hunk 11]
```diff
@@ -65,23 +81,85 @@ async function setupEnv(args, requirements, opts = {}) {
...
     await ensureWallet(args);
   }
 }
 
-async function ensureRunningNode() {
-  const chainName = process.env.FORGE_CURRENT_CHAIN;
-  if (!(await isForgeStarted())) {
-    shell.echo(`${symbols.error} chain ${chalk.yellow(chainName)} is not started yet!`);
-    shell.echo(`${symbols.info} Please run ${chalk.cyan(`forge start -c ${chainName}`)} first!`);
+async function ensureRunningNode(requirement) {
```
Comment (polunzh) [inline src/core/env.js]: runningChain

File: src/core/env.js [hunk 12]
```diff
@@ -65,23 +81,85 @@ async function setupEnv(args, requirements, opts = {}) {
...
+  if (args.chainName) {
+    process.env.FORGE_CURRENT_CHAIN = args.chainName;
+    return;
+  }
+
+  if (!requirement) {
+    return;
+  }
+
+  const chainNames = getAllChainNames();
```
Comment (polunzh) [inline src/core/env.js]: common有封装的方法。

File: src/core/env.js [hunk 13]
```diff
@@ -65,23 +81,85 @@ async function setupEnv(args, requirements, opts = {}) {
...
+  if (chainNames.length === 0) {
+    printWarning(`There is no chains, please create it by run ${chalk.cyan('forge chain:create')}`);
+    process.exit(1);
+  }
+
+  if (requirement === true) {
+    process.env.FORGE_CURRENT_CHAIN = chainNames[0][0]; // eslint-disable-line
```
Comment (polunzh) [inline src/core/env.js]: 再封装一次getAllChainNames

File: src/core/env.js [hunk 14]
```diff
@@ -65,23 +81,85 @@ async function setupEnv(args, requirements, opts = {}) {
...
+}
+
+async function ensureCurrentChainRunning(requirement, chainName) {
+  if (!requirement) {
+    return true;
+  }
+
+  if (!(await isForgeStarted(chainName))) {
+    printWarning(`Chain ${chalk.yellow(chainName)} is not started yet!`);
+    printInfo(`Please run ${chalk.cyan(`forge start ${chainName}`)} first!`);
```
Comment (polunzh) [inline src/core/env.js:175]: warnning时process退出码为0

File: src/core/env.js [hunk 15]
```diff
@@ -129,14 +205,14 @@ function ensureRpcClient(args) {
       `${symbols.success} Using forge config: ${util.inspect(config, { depth: 5, colors: true })}`
     );
   } else {
-    shell.echo(`${symbols.error} forge-cli requires a valid forge config file to start
+    printError(`forge-cli requires a valid forge config file to start
```
Comment (polunzh) [inline src/core/env.js]: 直接throw 没有有效的配置

File: src/cli/node/config/lib.js [hunk 16]
```diff
@@ -458,7 +461,7 @@ async function readUserConfigs(
     tokenHolderPk,
   } = answers;
 
-  const chainId = strEqual(chainName, DEFAULT_CHAIN_NAME) ? 'forge' : chainName;
+  const chainId = strEqual(name, DEFAULT_CHAIN_NAME) ? 'forge' : name;
```
Comment (polunzh) [inline src/cli/node/config/lib.js]: 删除 DEFAULT_CHAIN_NAME

File: src/cli/node/stop/index.js (hunk 6 above)
Comment (polunzh) [reply src/cli/node/stop/index.js]: 这个地方不能用 `requirement.currentChainRunning` 替代，因为这个地方是取当前启动的链，而不是看当前链是不是启动。

**Comment by wangshijun** [comment]
Bug:

```
❯ ls ~/.forge_chains
ls: /Users/wangshijun/.forge_chains: No such file or directory
```

```
❯ forge chain:create wangshijun -d
✔ Initialized an empty storage space in /Users/wangshijun/.forge_chains/forge_default
✔ Extract forge config from /Users/wangshijun/.forge_cli/release/forge/0.38.2/lib/forge-0.38.2/priv/forge_release.toml
✔ Forge config written to /Users/wangshijun/.forge_chains/forge_default/forge_release.toml
✔ chain name: wangshijun
✖ Create new chain failed:
✖ Cannot read property 'address' of undefined
```

**Comment by wangshijun** [comment]
@polunzh please run test with a clean setup (empty .forge_cli and empty .forge_chains)

- `forge install`
- `forge chain:create xxx -d`
- `forge start`
- `forge status`
- `forge stop`
- `forge download`
- `forge upgrade`

As I have tested, the second step is not robust due to following reasons:

- the chain:create will crash if `FORGE_MODERATOR_SK` not set in env
- neither `forge config set` nor `forge chain:create` prints the randomly generated token holder wallet on complete.

## duplicate chain creating issue

cli fails without any error message or error log when I create a chain with name that has already been created.

```
forge chain:create wangshijun -d
forge chain:create wangshijun -d
```

## `wallet:*` should not require a chain

```
❯ forge wallet:create --defaults
⚠ There are no chains, please create it by run forge chain:create
```

But `account:*` requires a chain

## web port for first chain should be 8210?

Only 1 chain was ever created on my test machine, but the web port is not 8210, I strongly suggest we keep things consistent, because 8210 is a widely known port for our developer users.

**Review by wangshijun — CHANGES_REQUESTED** [review]
I have some comments in the pr, please resolve them before publish

**Comment by polunzh** [comment]
> @polunzh please run test with a clean setup (empty .forge_cli and empty .forge_chains)
> 
> * `forge install`
> * `forge chain:create xxx -d`
> * `forge start`
> * `forge status`
> * `forge stop`
> * `forge download`
> * `forge upgrade`
> 
> As I have tested, the second step is not robust due to following reasons:
> 
> * the chain:create will crash if `FORGE_MODERATOR_SK` not set in env
> * neither `forge config set` nor `forge chain:create` prints the randomly generated token holder wallet on complete.
> 
> ## duplicate chain creating issue
> cli fails without any error message or error log when I create a chain with name that has already been created.
> 
> ```
> forge chain:create wangshijun -d
> forge chain:create wangshijun -d
> ```
> 
> ## `wallet:*` should not require a chain
> ```
> ❯ forge wallet:create --defaults
> ⚠ There are no chains, please create it by run forge chain:create
> ```
> 
> But `account:*` requires a chain
> 
> ## web port for first chain should be 8210?
> Only 1 chain was ever created on my test machine, but the web port is not 8210, I strongly suggest we keep things consistent, because 8210 is a widely known port for our developer users.

Resolved:
1. fixed duplicate chain creating issue
1. `wallet:create` command is no longer need a chain
1. `account:*` commands need an existed chain
1.  first chain's default web port will be 8210
1. if the chain used the generated token holder, cli will print the holder's sk/pk/address

**Review by wangshijun — APPROVED** [review]

**Comment by polunzh** [comment]
Relate issue:
https://github.com/ArcBlock/forge-cli/issues/240

**Pull request merged**